### PR TITLE
Add fixes to Monads found by new Law checks

### PR DESCRIPTION
**Pull request by pakoito**
This PR adds some missing functions and equality check required for some monads to be lawful.

### CODE REVIEW

**Comment by pakoito** [comment]
Note that the Eq instance in the laws cannot be a default value because of this bug: https://youtrack.jetbrains.com/oauth?state=%2Fissue%2FKT-18660

File: kategory/src/main/kotlin/kategory/typeclasses/Eq.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package kategory
+
+interface Eq<in F> : Typeclass {
```
Comment (raulraja) [inline kategory/src/main/kotlin/kategory/typeclasses/Eq.kt]: :clap:

File: kategory/src/main/kotlin/kategory/typeclasses/Eq.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    fun eqv(a: F, b: F): Boolean
+
+    fun neqv(a: F, b: F): Boolean =
+            !eqv(a, b)
+
+    companion object {
+        operator fun <F> invoke() = object : Eq<F> {
```
Comment (raulraja) [inline kategory/src/main/kotlin/kategory/typeclasses/Eq.kt]: I think this should be a default instance of `Eq[Any]` outside of this object.

File: kategory/src/test/kotlin/kategory/data/EitherTTest.kt [hunk 3]
```diff
@@ -10,7 +10,7 @@ import org.junit.runner.RunWith
 class EitherTTest : UnitSpec() {
     init {
 
-        testLaws(MonadLaws.laws(EitherTMonad<Id.F, Int>()))
+        testLaws(MonadLaws.laws(EitherTMonad<Id.F, Int>(), Eq()))
```
Comment (raulraja) [inline kategory/src/test/kotlin/kategory/data/EitherTTest.kt]: Probably better named as `EqAny`

File: kategory/src/test/kotlin/kategory/data/EvalTest.kt [hunk 4]
```diff
@@ -8,7 +8,10 @@ import org.junit.runner.RunWith
 class EvalTest : UnitSpec() {
     init {
 
-        testLaws(MonadLaws.laws(Eval))
+        testLaws(MonadLaws.laws(Eval, object : Eq<HK<Eval.F, Int>> {
```
Comment (raulraja) [inline kategory/src/test/kotlin/kategory/data/EvalTest.kt]: We should have this instance in the core not just in tests and there should be an instance for each one of the datatypes we implement.

File: kategory/src/test/kotlin/kategory/data/Function0Test.kt [hunk 5]
```diff
@@ -8,7 +8,10 @@ import org.junit.runner.RunWith
 class Function0Test : UnitSpec() {
     init {
 
-        testLaws(MonadLaws.laws(Function0))
+        testLaws(MonadLaws.laws(Function0, object : Eq<HK<Function0.F, Int>> {
```
Comment (raulraja) [inline kategory/src/test/kotlin/kategory/data/Function0Test.kt]: We should have this instance in the core not just in tests.

File: kategory/src/test/kotlin/kategory/laws/Law.kt [hunk 6]
```diff
@@ -2,3 +2,5 @@ package kategory
 
 data class Law(val name: String, val test: () -> Unit)
 
+inline fun <reified A> A.equalUnderTheLaw(b: A, eq: Eq<A> = Eq()): Boolean =
```
Comment (raulraja) [inline kategory/src/test/kotlin/kategory/laws/Law.kt]: Can we have extension functions that are `inline reify... <F, ..>` Once we have instances for the datatypes as part of our effort to provide syntax all datatypes will include syntax for `eqv` and `neqv` which can be used directly here provided thre is a registeres `Eq` instance for the datatype being compared.

File: kategory/src/test/kotlin/kategory/laws/MonadLaws.kt [hunk 7]
```diff
@@ -1,49 +1,47 @@
...
 object MonadLaws {
 
-    inline fun <reified F> laws(M: Monad<F> = monad<F>()): List<Law> =
-            ApplicativeLaws.laws(M) + listOf(
-                    Law("Monad Laws: left identity", { leftIdentity(M) }),
-                    Law("Monad Laws: right identity", { rightIdentity(M) }),
-                    Law("Monad Laws: kleisli left identity", { kleisliLeftIdentity(M) }),
-                    Law("Monad Laws: kleisli right identity", { kleisliRightIdentity(M) }),
-                    Law("Monad Laws: map / flatMap coherence", { mapFlatMapCoherence(M) }),
-                    Law("Monad / JVM: stack safe", { mapFlatMapCoherence(M) })
```
Comment (raulraja) [inline kategory/src/test/kotlin/kategory/laws/MonadLaws.kt]: Thanks!

**Pull request merged**